tests/linalg_test.py

```diff
@@ -973,6 +993,33 @@ def testLuGrad(self, shape, dtype, rng_factory):
     lu = vmap(jsp.linalg.lu) if len(shape) > 2 else jsp.linalg.lu
     jtu.check_grads(lu, (a,), 2, atol=5e-2, rtol=3e-1)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_corank={}".format(corank), "corank": corank}
+      for corank in [1, 2]))
+  @jtu.skip_on_devices("tpu")  # TODO(mattjj, pfau): fails on TPU.
+  def testLuGradOfSingularMatrix(self, corank):
+    def _lu(a):
+      p, l, u = jsp.linalg.lu(a)
+      return p, l[:, :-corank], u
+    mat = singular_mats[corank-1]
+    # Check forward mode gradients
+    jtu.check_grads(_lu, (mat,), 2, modes=['fwd'], atol=1e-1, rtol=1e-1)
+    # Something is causing tests to fail for reverse-mode.
```

Comment:
I suspect that the automated reverse mode check fails because it passes in a cotangent with nonzero elements outside of the triangular matrix.

Can you instead make the `_lu` helper return only the lower/upper triangular parts of the results? I.e., the part that is allowed to take on nonzero values?

Reply:
I don't think this is the issue, since `jsp.linalg.lu` masks out the upper and lower triangulars with zero anyway, and `jacfwd` and `jacrev` match.

Reply:
The test that's here now - looking at the entire Jacobian in forward and reverse mode, instead of just a random projection - is actually a stronger test that `jtu.check_grads`. So I'm inclined to leave it.

Reply:
Hmm, I just tried this out and you may be right.

I'm a little nervous that we aren't testing higher order derivatives in reverse mode, but this is definitely an improvement.